### PR TITLE
Admin merchants

### PR DESCRIPTION
**Pull request by Scott-Borecki**
### What I Did:
- Completed the following User Stories:
  - #11 
  - #12 
  - #13 
  - #14 
  - #15 
  - #16 
  - #17 
- Added through relationships in the models and updated the relationship tests
- Added `enabled` column to Merchant with a default value of `false`

### Road Blocks
- Setting up tests with FactoryBot when all the models were needed.
- The ActiveRecord queries ... 😅😬
- I tried building helper methods for the ActiveRecord queries, but it didn't seem to ... help.

### Test Coverage
- [X] I have tested this code with RSpec
- All the tests pass, but need to update the objects so that there are invoices that have items from different merchants ... so we can test all the edge cases and make sure functionality is working like we expect.
  - In particular, `#top_five_merchants_by_revenue` might not be calculating correctly.
- `#best_day` method and testing is incomplete.

### Questions/Notes:
- Need to improve test setup by making more combinations of objects.
- There will probably be some merge conflicts we'll need to deal with after merging Kim's branch.
- We'll need to confirm tests are working now that there is a new `enabled` column in the `Merchant` table.
- I would like to refactor my factories.  There has to be a cleaner way to do that with create_list and nesting factories.

### CODE REVIEW

File: app/controllers/admin/merchants_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  def new
+  end
+
+  def create
+    # TODO (Scott Borecki): Add sad path if creation fails
+    Merchant.create!(merchant_params)
+    redirect_to admin_merchants_path
+  end
+
+  def edit
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply app/controllers/admin/merchants_controller.rb:21]: Ignore this comment for controller actions

File: app/controllers/admin/merchants_controller.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  before_action :fetch_merchant_collection, only: [:index]
+  before_action :fetch_new_merchant, only: [:new]
+
+  def index
+  end
+
+  def show
+  end
+
+  def new
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply app/controllers/admin/merchants_controller.rb:12]: Ignore this comment for controller actions

File: app/controllers/admin/merchants_controller.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
+class Admin::MerchantsController < ApplicationController
+  before_action :fetch_current_merchant, only: [:edit, :show, :update]
+  before_action :fetch_merchant_collection, only: [:index]
+  before_action :fetch_new_merchant, only: [:new]
+
+  def index
+  end
+
+  def show
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply app/controllers/admin/merchants_controller.rb:9]: Ignore this comment for controller actions

File: app/controllers/admin/merchants_controller.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
+class Admin::MerchantsController < ApplicationController
+  before_action :fetch_current_merchant, only: [:edit, :show, :update]
+  before_action :fetch_merchant_collection, only: [:index]
+  before_action :fetch_new_merchant, only: [:new]
+
+  def index
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply app/controllers/admin/merchants_controller.rb:6]: Ignore this comment for controller actions

File: app/controllers/admin/merchants_controller.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
+class Admin::MerchantsController < ApplicationController
+  before_action :fetch_current_merchant, only: [:edit, :show, :update]
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply app/controllers/admin/merchants_controller.rb:2]: Ignore this comment for controllers

File: config/routes.rb [hunk 6]
```diff
@@ -1,4 +1,7 @@
 Rails.application.routes.draw do
   get '/', to: 'application#welcome'
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  namespace :admin do
+    resources :merchants, only: [:create, :edit, :index, :new, :show, :update]
```
[the remark this answers is not in the document]
Comment (kabcouwer) [reply config/routes.rb:5]: nice I like the look of this a lot - I'd be down to use this syntax with my routes when we merge our branches

File: app/controllers/admin/merchants_controller.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  before_action :fetch_merchant_collection, only: [:index]
+  before_action :fetch_new_merchant, only: [:new]
```
Comment (kabcouwer) [inline app/controllers/admin/merchants_controller.rb:4]: curious what these lines of code are doing 😎

File: config/routes.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply config/routes.rb:5]: Ignore for routes

File: db/schema.rb [hunk 8]
```diff
@@ -57,6 +57,7 @@
     t.string "name"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.boolean "enabled", default: false, null: false
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply db/schema.rb:60]: Ignore Hound comments on schema

File: db/schema.rb [hunk 9]
```diff
@@ -10,7 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2021_07_26_225210) do
+ActiveRecord::Schema.define(version: 2021_07_29_025435) do
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply db/schema.rb:13]: Ignore Hound comments on schema

File: db/migrate/20210729025435_change_enabled_default_to_be_false_in_merchants.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class ChangeEnabledDefaultToBeFalseInMerchants < ActiveRecord::Migration[5.2]
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply db/migrate/20210729025435_change_enabled_default_to_be_false_in_merchants.rb:1]: Ignore Hound comments on migrations

File: db/migrate/20210729012753_add_enabled_to_merchants.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+class AddEnabledToMerchants < ActiveRecord::Migration[5.2]
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply db/migrate/20210729012753_add_enabled_to_merchants.rb:1]: Ignore Hound comments on migrations

File: spec/features/admin_merchants/edit_spec.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+          expect(current_path).to eq(admin_merchant_path(merchant1))
+        end
+
+        it 'displays the updated information' do
+          expect(page).to have_content('Stompy Feet')
+          expect(page).to have_no_content(merchant2.name)
+          expect(page).to have_no_content(merchant3.name)
+        end
+
+        it 'displays a flash message stating that the information has been successfully updated' do
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/features/admin_merchants/edit_spec.rb:40]: Ignore line length comments in spec files

File: spec/features/admin_merchants/edit_spec.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  specify { expect(Merchant.all.count).to be_positive }
+
+  describe 'as an admin' do
+    describe 'when I visit the admin merchants edit page (/admin/merchants/merchant_id/edit)' do
+      before { visit edit_admin_merchant_path(merchant1) }
+
+      specify { expect(current_path).to eq(edit_admin_merchant_path(merchant1)) }
+      specify { expect(page).to have_no_content('Update successful!') }
+
+      it 'displays a form filled in with the existing merchant attribute information' do
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/features/admin_merchants/edit_spec.rb:17]: Ignore line length comments in spec files

File: spec/features/admin_merchants/edit_spec.rb [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  let!(:merchant2) { create(:enabled_merchant) }
+  let!(:merchant3) { create(:enabled_merchant) }
+
+  specify { expect(Merchant.all.count).to be_positive }
+
+  describe 'as an admin' do
+    describe 'when I visit the admin merchants edit page (/admin/merchants/merchant_id/edit)' do
+      before { visit edit_admin_merchant_path(merchant1) }
+
+      specify { expect(current_path).to eq(edit_admin_merchant_path(merchant1)) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/features/admin_merchants/edit_spec.rb:14]: Ignore line length comments in spec files

File: spec/features/admin_merchants/edit_spec.rb [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+RSpec.describe '/admin/merchants/merchant_id/edit' do
+  let!(:merchant1) { create(:enabled_merchant) }
+  let!(:merchant2) { create(:enabled_merchant) }
+  let!(:merchant3) { create(:enabled_merchant) }
+
+  specify { expect(Merchant.all.count).to be_positive }
+
+  describe 'as an admin' do
+    describe 'when I visit the admin merchants edit page (/admin/merchants/merchant_id/edit)' do
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/features/admin_merchants/edit_spec.rb:11]: Ignore line length comments in spec files

File: spec/features/admin_merchants/edit_spec.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+RSpec.describe '/admin/merchants/merchant_id/edit' do
+  let!(:merchant1) { create(:enabled_merchant) }
+  let!(:merchant2) { create(:enabled_merchant) }
+  let!(:merchant3) { create(:enabled_merchant) }
+
+  specify { expect(Merchant.all.count).to be_positive }
+
+  describe 'as an admin' do
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/features/admin_merchants/edit_spec.rb:10]: Ignore line length comments in spec files

File: spec/features/admin_merchants/edit_spec.rb [hunk 17]
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+RSpec.describe '/admin/merchants/merchant_id/edit' do
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/features/admin_merchants/edit_spec.rb:3]: Ignore line length comments in spec files

File: app/models/merchant.rb [hunk 18]
```diff
@@ -1,5 +1,72 @@
...
+    select('merchants.*',
+           'SUM(invoice_items.quantity * invoice_items.unit_price) AS revenue')
+      .joins(:transactions)
+      .where(transactions: { result: :success })
+      .group('merchants.id')
+      .order('revenue desc')
+      .limit(5)
+  end
+
+  def self.total_revenue_generated_by_merchant(merchant)
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply app/models/merchant.rb:36]: We can see if we can refactor this later

File: spec/factories.rb [hunk 19]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
+  let!(:invoice_item3a) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 1, unit_price: 10) }
+  let!(:invoice_item3b) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 5, unit_price: 20) }
+  let!(:invoice_item4a) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 3, unit_price: 10) }
+  let!(:invoice_item4b) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 5, unit_price: 20) }
+  let!(:invoice_item5a) { create(:invoice_item, :shipped, item: item5, invoice: invoice5, quantity: 5, unit_price: 10) }
+  let!(:invoice_item5b) { create(:invoice_item, :shipped, item: item5, invoice: invoice5, quantity: 5, unit_price: 20) }
+  let!(:invoice_item6a) { create(:invoice_item, :shipped, item: item6, invoice: invoice6, quantity: 6, unit_price: 10) }
+  let!(:invoice_item6b) { create(:invoice_item, :shipped, item: item6, invoice: invoice6, quantity: 5, unit_price: 20) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:114]: Ignore

File: app/models/merchant.rb [hunk 20]
```diff
@@ -1,5 +1,62 @@
...
+
+  def self.disabled_merchants
+    where(enabled: false)
+  end
+
+  def enabled?
+    enabled
+  end
+
+  def self.top_five_merchants_by_revenue
```
Comment (kabcouwer) [inline app/models/merchant.rb:26]: nice method!!

File: spec/factories.rb [hunk 21]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
+  let!(:invoice_item3a) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 1, unit_price: 10) }
+  let!(:invoice_item3b) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 5, unit_price: 20) }
+  let!(:invoice_item4a) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 3, unit_price: 10) }
+  let!(:invoice_item4b) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 5, unit_price: 20) }
+  let!(:invoice_item5a) { create(:invoice_item, :shipped, item: item5, invoice: invoice5, quantity: 5, unit_price: 10) }
+  let!(:invoice_item5b) { create(:invoice_item, :shipped, item: item5, invoice: invoice5, quantity: 5, unit_price: 20) }
+  let!(:invoice_item6a) { create(:invoice_item, :shipped, item: item6, invoice: invoice6, quantity: 6, unit_price: 10) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:113]: Ignore

File: spec/factories.rb [hunk 22]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
+  let!(:invoice_item3a) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 1, unit_price: 10) }
+  let!(:invoice_item3b) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 5, unit_price: 20) }
+  let!(:invoice_item4a) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 3, unit_price: 10) }
+  let!(:invoice_item4b) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 5, unit_price: 20) }
+  let!(:invoice_item5a) { create(:invoice_item, :shipped, item: item5, invoice: invoice5, quantity: 5, unit_price: 10) }
+  let!(:invoice_item5b) { create(:invoice_item, :shipped, item: item5, invoice: invoice5, quantity: 5, unit_price: 20) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:112]: Ignore

File: spec/factories.rb [hunk 23]
```diff
@@ -1,13 +1,117 @@
...
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
+  let!(:invoice_item3a) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 1, unit_price: 10) }
+  let!(:invoice_item3b) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 5, unit_price: 20) }
+  let!(:invoice_item4a) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 3, unit_price: 10) }
+  let!(:invoice_item4b) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 5, unit_price: 20) }
+  let!(:invoice_item5a) { create(:invoice_item, :shipped, item: item5, invoice: invoice5, quantity: 5, unit_price: 10) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:111]: Ignore

File: spec/factories.rb [hunk 24]
```diff
@@ -1,13 +1,117 @@
...
+
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
+  let!(:invoice_item3a) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 1, unit_price: 10) }
+  let!(:invoice_item3b) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 5, unit_price: 20) }
+  let!(:invoice_item4a) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 3, unit_price: 10) }
+  let!(:invoice_item4b) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 5, unit_price: 20) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:110]: Ignore

File: spec/factories.rb [hunk 25]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:transaction6) { create(:transaction, :success, invoice: invoice6) }
+
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
+  let!(:invoice_item3a) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 1, unit_price: 10) }
+  let!(:invoice_item3b) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 5, unit_price: 20) }
+  let!(:invoice_item4a) { create(:invoice_item, :shipped, item: item4, invoice: invoice4, quantity: 3, unit_price: 10) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:109]: Ignore

File: spec/factories.rb [hunk 26]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:transaction5) { create(:transaction, :success, invoice: invoice5) }
+  let!(:transaction6) { create(:transaction, :success, invoice: invoice6) }
+
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
+  let!(:invoice_item3a) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 1, unit_price: 10) }
+  let!(:invoice_item3b) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 5, unit_price: 20) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:108]: Ignore

File: spec/factories.rb [hunk 27]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:transaction4) { create(:transaction, :success, invoice: invoice4) }
+  let!(:transaction5) { create(:transaction, :success, invoice: invoice5) }
+  let!(:transaction6) { create(:transaction, :success, invoice: invoice6) }
+
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
+  let!(:invoice_item3a) { create(:invoice_item, :shipped, item: item3, invoice: invoice3, quantity: 1, unit_price: 10) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:107]: Ignore

**Comment by kabcouwer** [comment]
Some of these methods are very impressive Scott! nice work! I need to get ready for work but I'm gonna continue to look through your code, but from what I've seen, looks great so far!

File: app/controllers/admin/merchants_controller.rb (hunk 7 above)
Comment (Scott-Borecki) [reply app/controllers/admin/merchants_controller.rb:4]: For sure!  Basically what those are doing is calling the methods listed in the private methods at the bottom.  Then those methods get called with the actions in the array after the `only:`.  

In this case, it makes the most sense for the `fetch_current_merchant` method since the same instance variable is used in each of the three actions and it DRYs up the code ... instead of having to repeat the same line in each action that stores the Merchant object in the instance variable.  You could argue it is not really DRYing up the code for the `fetch_merchant_collection` and `fetch_new_merchant` methods (...well it isn't DRYing it up at all here! haha!), but I just went with it to be consistent.

You can check out a little more on it here:
- https://guides.rubyonrails.org/v5.2/action_controller_overview.html#filters

File: spec/factories.rb [hunk 28]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:transaction3) { create(:transaction, :success, invoice: invoice3) }
+  let!(:transaction4) { create(:transaction, :success, invoice: invoice4) }
+  let!(:transaction5) { create(:transaction, :success, invoice: invoice5) }
+  let!(:transaction6) { create(:transaction, :success, invoice: invoice6) }
+
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
+  let!(:invoice_item2b) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 5, unit_price: 20) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:106]: Ignore

File: spec/factories.rb [hunk 29]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:transaction2) { create(:transaction, :success, invoice: invoice2) }
+  let!(:transaction3) { create(:transaction, :success, invoice: invoice3) }
+  let!(:transaction4) { create(:transaction, :success, invoice: invoice4) }
+  let!(:transaction5) { create(:transaction, :success, invoice: invoice5) }
+  let!(:transaction6) { create(:transaction, :success, invoice: invoice6) }
+
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
+  let!(:invoice_item2a) { create(:invoice_item, :shipped, item: item2, invoice: invoice2, quantity: 4, unit_price: 10) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:105]: Ignore

File: spec/factories.rb [hunk 30]
```diff
@@ -1,13 +1,117 @@
...
+  let!(:transaction1) { create(:transaction, :failed, invoice: invoice1) }
+  let!(:transaction2) { create(:transaction, :success, invoice: invoice2) }
+  let!(:transaction3) { create(:transaction, :success, invoice: invoice3) }
+  let!(:transaction4) { create(:transaction, :success, invoice: invoice4) }
+  let!(:transaction5) { create(:transaction, :success, invoice: invoice5) }
+  let!(:transaction6) { create(:transaction, :success, invoice: invoice6) }
+
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
+  let!(:invoice_item1b) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 5, unit_price: 20) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:104]: Ignore

File: spec/factories.rb [hunk 31]
```diff
@@ -1,13 +1,117 @@
...
+  # TRANSACTIONS
+  let!(:transaction1) { create(:transaction, :failed, invoice: invoice1) }
+  let!(:transaction2) { create(:transaction, :success, invoice: invoice2) }
+  let!(:transaction3) { create(:transaction, :success, invoice: invoice3) }
+  let!(:transaction4) { create(:transaction, :success, invoice: invoice4) }
+  let!(:transaction5) { create(:transaction, :success, invoice: invoice5) }
+  let!(:transaction6) { create(:transaction, :success, invoice: invoice6) }
+
+  # INVOICE ITEMS
+  let!(:invoice_item1a) { create(:invoice_item, :shipped, item: item1, invoice: invoice1, quantity: 2, unit_price: 10) }
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:103]: Ignore

File: spec/factories.rb [hunk 32]
```diff
@@ -1,13 +1,117 @@
...
+
+  # Transaction
+  factory :transaction do
+    credit_card_number { Faker::Business.credit_card_number.delete('-') }
+    credit_card_expiration_date { Faker::Business.credit_card_expiry_date }
+    invoice
+  end
+end
+
+def create_factories
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:66]: Ignore

File: spec/factories.rb [hunk 33]
```diff
@@ -1,13 +1,117 @@
...
+    #   factory :merchant_with_revenue_generated do
+    #     transient do
+    #       items_count { 5 }
+    #     end
+    #
+    #     after(:create) do |merchant, evaluator|
+    #       create_list(:item, evaluator.items_count, merchant: merchant)
+    #       # merchant.reload
+    #     end
+    #   end
```
[the remark this answers is not in the document]
Comment (Scott-Borecki) [reply spec/factories.rb:50]: Ignore

**Comment by AlexKlick** [comment]
I merged Scott's PR into Kim's. There was a duplication of instance method variable checks, but this shouldn't be an issue. Can clean that up as a team, as I don't want to accidentally delete something that should be there.
